_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
+# attributes: Count, 99thPercentile
+```
+
+Likewise the following producer metrics can be helpful, perhaps after broker logs reveal some misbehavior.
+
+```sh
+kafka.producer:type=producer-metrics,client-id=([-.\w]+)
+# attributes: txn-init-time-ns-total, txn-begin-time-ns-total, txn-send-offsets-time-ns-total, txn-commit-time-ns-total, txn-abort-time-ns-total
+```
+
+### Drawbacks
+
+There are a few drawbacks when using Kafka transactions that developers should be aware of.
+
+Firstly, enabling EOS has an effect on throughput, which can be significant.
+Each producer can only process transactions sequentially, so you might need multiple transactional producers, which can have a knock-on effect on the rest of the application.
+This is where frameworks can provide useful abstractions over multiple producer instances.
+
+Then, transactions can add some latency to an application due to the write amplification required to ensure EOS.
+While this may not be a significant issue for some applications, those that require real-time processing may be affected.
+
+Each transaction in Kafka incurs some overhead that is independent of the number of messages involved. 
+This overhead includes the following:
+
+- A small number of Remote Procedure Calls (RPCs) to register the partitions with the coordinator. 
+  These RPCs can be batched together to reduce overhead.
+- A marker that must be written to each partition when the transaction is completed.
+- A few state change records that are appended to the internal transaction state log.
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.
+
+Another potential drawback is hanging transactions, which are transactions with a missing or out-of-order control record.
+They may be caused by a network issue causing delayed messages ([KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).
+This issue is rare, but it's important to be aware and set alerts, because it can impact service availability.
+
+A hanging transaction is usually revealed by a stuck application with growing lag ([KIP-664](https://cwiki.apache.org/confluence/display/KAFKA/KIP-664%3A+Provide+tooling+to+detect+and+abort+hanging+transactions)).
+Despite messages being produced, consumers can't make any progress because the LSO does not grow anymore.
```

Comment:
You should mention here about transaction timeout, and how hanging transactions are different.

Reply:
I would say:

The transaction coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
This mechanism does not work for hanging transactions because they are marked as completed.